File: .skills/openclaw-skills/skills/asif2bd/openclaw-tour-planner/src/apis/weather.ts
import axios from 'axios';
import { WeatherForecast, WeatherDay } from '../types';

/**
 * WMO Weather interpretation codes → human-readable conditions
 * https://open-meteo.com/en/docs#weathervariables
 */
const WMO_CODES: Record<number, string> = {
  0: 'Clear sky', 1: 'Mainly clear', 2: 'Partly cloudy', 3: 'Overcast',
  45: 'Fog', 48: 'Icy fog',
  51: 'Light drizzle', 53: 'Moderate drizzle', 55: 'Dense drizzle',
  61: 'Light rain', 63: 'Moderate rain', 65: 'Heavy rain',
  71: 'Light snow', 73: 'Moderate snow', 75: 'Heavy snow',
  77: 'Snow grains',
  80: 'Light showers', 81: 'Moderate showers', 82: 'Violent showers',
  85: 'Snow showers', 86: 'Heavy snow showers',
  95: 'Thunderstorm', 96: 'Thunderstorm with hail', 99: 'Thunderstorm with heavy hail',
};

function wmoToCondition(code: number): string {
  return WMO_CODES[code] ?? 'Unknown';
}

/**
 * Open-Meteo weather client — free, no API key required
 * https://api.open-meteo.com
 */
export class WeatherClient {
  private readonly openMeteoBase = 'https://api.open-meteo.com/v1/forecast';
  private readonly visualCrossingBase =
    'https://weather.visualcrossing.com/VisualCrossingWebServices/rest/services/timeline';

  /**
   * Get weather forecast for coordinates
   * Primary: Open-Meteo (free, no key)
   * Fallback: Visual Crossing (requires VISUAL_CROSSING_API_KEY env var)
   */
  async getForecast(
    destinationName: string,
    lat: number,
    lon: number,
    days: number = 7,
  ): Promise<WeatherForecast> {
    try {
      return await this.fetchOpenMeteo(destinationName, lat, lon, days);
    } catch (primaryErr) {
      const vcKey = process.env.VISUAL_CROSSING_API_KEY;
      if (vcKey) {
        try {
          return await this.fetchVisualCrossing(destinationName, lat, lon, days, vcKey);
        } catch {
          // fall through to error
        }
      }
      throw new Error(
        `Weather data unavailable for ${destinationName}. ` +
          `Set VISUAL_CROSSING_API_KEY for a fallback source. (${(primaryErr as Error).message})`,
      );
    }
  }

  private async fetchOpenMeteo(
    destinationName: string,
    lat: number,
    lon: number,
    days: number,
  ): Promise<WeatherForecast> {
    const response = await axios.get(this.openMeteoBase, {
      params: {
        latitude: lat,
        longitude: lon,
        daily: [
          'weathercode',
          'temperature_2m_max',
          'temperature_2m_min',
          'precipitation_probability_max',
          'windspeed_10m_max',
          'relative_humidity_2m_max',
        ].join(','),
        timezone: 'auto',
        forecast_days: Math.min(days, 16),
      },
      timeout: 8000,
    });

    const d = response.data.daily;
    const weatherDays: WeatherDay[] = d.time.map((date: string, i: number) => ({
      date,
      tempMax: Math.round(d.temperature_2m_max[i]),
      tempMin: Math.round(d.temperature_2m_min[i]),
      conditions: wmoToCondition(d.weathercode[i]),
      precipitationChance: d.precipitation_probability_max[i] ?? 0,
      humidity: d.relative_humidity_2m_max[i] ?? 0,
      windSpeed: Math.round(d.windspeed_10m_max[i]),
    }));

    return {
      destination: destinationName,
      days: weatherDays,
      summary: this.buildSummary(weatherDays),
    };
  }

  private async fetchVisualCrossing(
    destinationName: string,
    lat: number,
    lon: number,
    days: number,
    apiKey: string,
  ): Promise<WeatherForecast> {
    const url = `${this.visualCrossingBase}/${lat},${lon}/next${days}days`;
    const response = await axios.get(url, {
      params: { unitGroup: 'metric', key: apiKey, contentType: 'json' },
      timeout: 8000,
    });

    const weatherDays: WeatherDay[] = response.data.days
      .slice(0, days)
      .map((day: any) => ({
        date: day.datetime,
        tempMax: Math.round(day.tempmax),
        tempMin: Math.round(day.tempmin),
        conditions: day.conditions ?? 'Unknown',
        precipitationChance: Math.round((day.precipprob ?? 0)),
        humidity: Math.round(day.humidity ?? 0),
        windSpeed: Math.round(day.windspeed ?? 0),
      }));

    return {
      destination: destinationName,
      days: weatherDays,
      summary: this.buildSummary(weatherDays),
    };
  }

  private buildSummary(days: WeatherDay[]): string {
    if (!days.length) return 'No forecast available.';
    const avgHigh = Math.round(days.reduce((s, d) => s + d.tempMax, 0) / days.length);
    const avgLow  = Math.round(days.reduce((s, d) => s + d.tempMin, 0) / days.length);
    const rainyDays = days.filter(d => d.precipitationChance > 40).length;
    const conditions = days[0]?.conditions ?? 'variable';
    return (
      `${avgLow}–${avgHigh}°C, ${conditions.toLowerCase()}` +
      (rainyDays > 0 ? `, ${rainyDays} day(s) with rain likely` : ', low rain chance')
    );
  }
}

export const weatherClient = new WeatherClient();

File: .skills/openclaw-skills/skills/timo2026/travel-swarm/backend/utils/weather.py
"""天气查询模块 - 按时间节点查询天气预报"""
import aiohttp
import os

class WeatherClient:
    """天气查询客户端（wttr.in / Open-Meteo）"""
    
    def __init__(self):
        self.api_url = "https://wttr.in"
    
    async def get_weather(self, city: str, date: str = "") -> dict:
        """查询天气预报
        
        Args:
            city: 城市
            date: 日期（如"2026-05-01"）
        
        Returns:
            天气信息
        """
        
        print(f"[天气] 查询: {city} 日期:{date}")
        
        # wttr.in 格式查询
        try:
            async with aiohttp.ClientSession() as session:
                url = f"{self.api_url}/{city}?format=j1"
                async with session.get(url, timeout=10) as resp:
                    if resp.status == 200:
                        data = await resp.json()
                        return self._parse_weather(data, date)
        except Exception as e:
            print(f"[天气] API失败: {e}")
        
        # 备用：生成模拟数据
        return self._generate_mock_weather(city, date)
    
    def _parse_weather(self, data: dict, date: str) -> dict:
        """解析wttr.in返回数据"""
        try:
            current = data.get("current_condition", [{}])[0]
            weather_desc = current.get("weatherDesc", [{"value": "晴"}])[0].get("value", "晴")
            temp = current.get("temp_C", "20")
            
            return {
                "city": data.get("area", [{}])[0].get("areaName", [{}])[0].get("value", "城市"),
                "date": date,
                "weather": weather_desc,
                "temp": f"{temp}°C",
                "humidity": current.get("humidity", "50%"),
                "wind": current.get("windspeedKmph", "10") + "km/h",
                "icon": self._get_weather_icon(weather_desc),
                "forecast": f"{weather_desc}为主"  # ⭐ 修复：添加forecast
            }
        except:
            return self._generate_mock_weather("城市", date)
    
    def _generate_mock_weather(self, city: str, date: str) -> dict:
        """生成模拟天气数据"""
        
        weather_by_city = {
            "西安": {"weather": "晴", "temp": "18°C", "icon": "☀️"},
            "上海": {"weather": "多云", "temp": "22°C", "icon": "⛅"},
            "北京": {"weather": "晴", "temp": "20°C", "icon": "☀️"},
            "杭州": {"weather": "小雨", "temp": "19°C", "icon": "🌧️"},
            "成都": {"weather": "阴", "temp": "17°C", "icon": "☁️"},
            "重庆": {"weather": "多云", "temp": "21°C", "icon": "⛅"},
            "南京": {"weather": "晴", "temp": "23°C", "icon": "☀️"},
            "苏州": {"weather": "多云", "temp": "22°C", "icon": "⛅"},
            "武汉": {"weather": "晴", "temp": "24°C", "icon": "☀️"},
            "长沙": {"weather": "多云", "temp": "25°C", "icon": "⛅"},
        }
        
        w = weather_by_city.get(city, {"weather": "晴", "temp": "20°C", "icon": "☀️"})
        
        return {
            "city": city,
            "date": date,
            "weather": w["weather"],
            "temp": w["temp"],
            "humidity": "60%",
            "wind": "15km/h",
            "icon": w["icon"],
            "forecast": f"{w['weather']}为主"
        }
    
    def _get_weather_icon(self, weather_desc: str) -> str:
        """根据天气描述获取图标"""
        icons = {
            "晴": "☀️", "Sunny": "☀️", "Clear": "☀️",
            "多云": "⛅", "Partly cloudy": "⛅",
            "阴": "☁️", "Overcast": "☁️",
            "小雨": "🌧️", "Light rain": "🌧️",
            "大雨": "⛈️", "Heavy rain": "⛈️",
            "雪": "❄️", "Snow": "❄️"
        }
        return icons.get(weather_desc, "🌤️")